### ISSUE TITLE
Sketcher: When multiple entities are overlayed, give user abilty to choose which entity to interact with

### ISSUE THREAD

**Issue by FreeCAD-Bug-Importer** [FreeCAD/FreeCAD#5648]
Issue imported from https://tracker.freecad.org/view.php?id=2492

* **Reporter:** mimosa
* **Date submitted:** 4/3/2016
* **FreeCAD version:** 0.15
* **Category:** Feature
* **Status:** new
* **Tags:** 

# Original report text

When more than one entity are overlayed (two lines, a line and the reference horizontal, 2 points, a point and the end point of a line, etc) the user should be able to chose which entity to interact with:

1) User would hover the mouse over (near) the area of interest,
2) If in that zone surrounding the mouse there exist more than one entity, after 2 second, a little button would appear (say, with 3 dots as a text)
3) when the button is clicked, a list would popup listing all the selectable entities in the vicinity
4) the user can select an entity (or close the popup without selecting any) and the program would continue (either the user was in the middle of an operation, ir want to start an operation with the selected entity).

# Other bug information

* **Priority:** normal
* **Severity:** feature
* **Category:** Feature
* **OS: Windows 10**
* **Platform:** PC
* **Updated:** 2/6/2021

# Discussion from Mantis ticket

### Comment by normandc 2016-04-05 00:34
There's already a mechanism to select overlapping sketch entities in v0.16 due to come out shortly.

---

### Comment by normandc 2018-01-06 19:16
abdullah wmayer

Should this be closed as fixed? In v0.17's Sketcher, we can now define the display order of normal/construction/reference geometry.

But this would not solve the issue where two lines of the same type are overlapping.

---

### Comment by abdullah 2018-01-11 10:13
normandc

I think it is an open point in the sketcher.

All you say is absolutely right, v.17 (probably also 0.16) has the elements widget which allows any selection. In this sense it is not a bug (but this ticket was opened as a feature). 

The new rendering order in v.17 will help as you comment... except in that case of two overlapping lines of the same type and one is fully overlapped by the other.

So let's say that the current implementation can be improved. 

Eventually, the whole inner parts of sketcher visualization (interface with coin3d) will need to be rewritten. It may be that in that effort this feature is "rescued".

**Comment by jthiesen** [comment]
I think #5867 solved this by allowing TAB to cycle between objects under cursor?

**Comment by adrianinsaval** [comment]
No, at most #5867 would be a duplicate of this, there it looks like it was solved for BIM workbench and Draft snap, but there isn't a general solution. I agree that TAB looks like a good button to use for this.

**Comment by luzpaz** [comment]
Looks like #8716 is related to this

**Comment by luzpaz** [comment]
this feels like a duplicate indeed. @PaddleStroke can you weigh in on this?

**Comment by PaddleStroke** [comment]
Yes this error is a subset of https://github.com/FreeCAD/FreeCAD/issues/5867.
Though perhaps sketcher will need a specific solution compared to the rest of freecad. I don't know.
I agree that using tab would be a good solution.

**Comment by karamellpelle** [comment]
Selecting things below should also be possible in the 3DView! Many times I need to select faces, edges, vertices etc that are visible behind transparent objects, but they are not selectable. Instead I  have to find and hide those obstacles in the TreeView, which is cumbersome.

In Blender you can <kbd>Alt</kbd> + <kbd>Left Click</kbd> to open a little menu when selecting overlapping objects. [Inkscape also uses <kbd>Alt</kbd> + <kbd>Left Click</kbd>](https://inkscape.org/doc/keys.html#id99) when selecting below.

<img width="547" height="290" alt="Image" src="https://github.com/user-attachments/assets/b7d3378b-0afd-4aa1-bc0d-00726c81681e" />

**Comment by leoheck** [comment]
It could use the same idea as Clarify Selection (improving clicking on cluttered areas).
Clarify Selection had not been implemented yet when this issue was created.

**Comment by Roy-043** [comment]
> It could use the same idea as Clarify Selection (improving clicking on cluttered areas).

The same was also suggested on the forum:
https://forum.freecad.org/viewtopic.php?t=105170

**Comment by Wallby** [comment]
Having one clarify selection tool helps for consistency. Legacy settings or tools that can be merged to make FreeCAD less intimidating should be a priority over adding yet another tool.
